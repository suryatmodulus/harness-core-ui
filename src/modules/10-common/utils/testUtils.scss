/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

.browser {
  border: 1px solid #cccccc;
  border-radius: 8px;
  min-height: 300px;
  overflow: auto;

  .browser-header {
    background-color: #cccccc;
    padding: 8px 16px;
    display: flex;
  }

  .browser-path {
    display: block;
    border: none;
    line-height: 14px;
    font-size: 12px;
    color: #444444;
    border-radius: 9px;
    min-width: 100px;
    padding: 2px 8px;
    flex: 1 1 auto;
    border: 1px solid transparent;

    &:focus {
      border-color: rgb(30, 167, 253);
      outline: none;
    }
  }

  .browser-content {
    padding: 8px;
  }
}
