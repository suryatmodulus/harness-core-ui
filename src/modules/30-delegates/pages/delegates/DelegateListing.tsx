/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useState, useEffect, useCallback, useMemo } from 'react'
import type { GetDataError } from 'restful-react'
import { get, set, pick, debounce } from 'lodash-es'
import type { FormikErrors } from 'formik'
import { useParams } from 'react-router-dom'
import {
  Container,
  Layout,
  FlexExpander,
  ExpandingSearchInput,
  useModalHook,
  SelectOption,
  FormInput,
  MultiSelectOption,
  PageError,
  shouldShowError
} from '@wings-software/uicore'
import { Dialog } from '@blueprintjs/core'
import { PageSpinner } from '@common/components'
import type { UseGetMockData } from '@common/utils/testUtils'
import FilterSelector from '@common/components/Filter/FilterSelector/FilterSelector'
import { Filter, FilterRef } from '@common/components/Filter/Filter'
import type { CrudOperation } from '@common/components/Filter/FilterCRUD/FilterCRUD'
import type { FilterInterface, FilterDataInterface } from '@common/components/Filter/Constants'
import {
  removeNullAndEmpty,
  isObjectEmpty,
  UNSAVED_FILTER,
  flattenObject
} from '@common/components/Filter/utils/FilterUtils'
import { ContainerSpinner } from '@common/components/ContainerSpinner/ContainerSpinner'

import { useStrings } from 'framework/strings'
import {
  GetDelegateGroupsNGV2WithFilterQueryParams,
  useGetDelegateGroupsNGV2WithFilter,
  DelegateGroupDetails
} from 'services/portal'
import { usePostFilter, useUpdateFilter, useDeleteFilter, useGetFilterList } from 'services/cd-ng'
import type { FilterDTO, ResponsePageFilterDTO, Failure, DelegateFilterProperties } from 'services/cd-ng'
import useCreateDelegateModal from '@delegates/modals/DelegateModal/useCreateDelegateModal'
import DelegateInstallationError from '@delegates/components/CreateDelegate/components/DelegateInstallationError/DelegateInstallationError'
import { useToaster, StringUtils } from '@common/exports'
import DelegatesEmptyState from '@delegates/images/DelegatesEmptyState.svg'
import RbacButton from '@rbac/components/Button/Button'
import { PermissionIdentifier } from '@rbac/interfaces/PermissionIdentifier'
import { ResourceType } from '@rbac/interfaces/ResourceType'
import DelegateListingItem, { DelegateListingHeader } from './DelegateListingItem'

import css from './DelegatesPage.module.scss'

const POLLING_INTERVAL = 10000

const statusTypes = ['ENABLED', 'WAITING_FOR_APPROVAL', 'DISABLED', 'DELETED']

const fullSizeContentStyle: React.CSSProperties = {
  width: '100%',
  height: '100%'
}

interface DelegatesListProps {
  filtersMockData?: UseGetMockData<ResponsePageFilterDTO>
}

export const DelegateListing: React.FC<DelegatesListProps> = ({ filtersMockData }) => {
  const { getString } = useStrings()
  const { accountId, projectIdentifier, orgIdentifier, module } = useParams<Record<string, string>>()
  const [searchTerm, setSearchTerm] = useState('')
  const [showDelegateLoader, setShowDelegateLoader] = useState<boolean>(true)
  const [page, setPage] = useState(0)
  const [filters, setFilters] = useState<FilterDTO[]>()
  const [appliedFilter, setAppliedFilter] = useState<FilterDTO | null>()
  const { showError } = useToaster()
  const [delegateGroups, setDelegateGroups] = useState<Array<DelegateGroupDetails>>([])
  const [delegateFetchError, setDelegateFetchError] = useState<GetDataError<Failure | Error>>()
  const [isRefreshingFilters, setIsRefreshingFilters] = useState<boolean>(false)
  const [isFilterOpen, setIsFilterOpen] = useState<boolean>(false)
  const filterRef = React.useRef<FilterRef<FilterDTO> | null>(null)

  const [troubleshoterOpen, setOpenTroubleshoter] = useState<{ isConnected: boolean | undefined } | undefined>(
    undefined
  )

  const queryParams: GetDelegateGroupsNGV2WithFilterQueryParams = useMemo(
    () =>
      ({
        accountId,
        orgId: orgIdentifier,
        projectId: projectIdentifier,
        module,
        pageIndex: page,
        pageSize: 10,
        searchTerm: ''
      } as GetDelegateGroupsNGV2WithFilterQueryParams),
    [accountId, module, orgIdentifier, projectIdentifier, page]
  )
  const { mutate: fetchDelegates, loading: isFetchingDelegates } = useGetDelegateGroupsNGV2WithFilter({ queryParams })
  const { openDelegateModal } = useCreateDelegateModal()

  useEffect(() => {
    setShowDelegateLoader(true)
    refetchDelegates(queryParams)
  }, [queryParams])

  const {
    loading: isFetchingFilters,
    data: fetchedFilterResponse,
    refetch: refetchFilterList
  } = useGetFilterList({
    queryParams: {
      accountIdentifier: accountId,
      projectIdentifier,
      orgIdentifier,
      type: 'Delegate'
    },
    mock: filtersMockData
  })

  useEffect(() => {
    setFilters(fetchedFilterResponse?.data?.content || [])
    setIsRefreshingFilters(isFetchingFilters)
  }, [fetchedFilterResponse])

  const getStatusSelectOptions = (values?: any[]): SelectOption[] => {
    const labelMap = {
      ENABLED: getString('enabledLabel'),
      DISABLED: getString('delegates.delGroupStatus.DISABLED'),
      WAITING_FOR_APPROVAL: getString('delegates.delGroupStatus.WAITING_FOR_APPROVAL'),
      DELETED: getString('deleted')
    }
    return values
      ? values.map(item => {
          return { label: get(labelMap, item, ''), value: item }
        })
      : []
  }

  const refetchDelegates = useCallback(
    async (params: GetDelegateGroupsNGV2WithFilterQueryParams, filter?): Promise<void> => {
      const { delegateGroupIdentifier, delegateName, delegateType, description, hostName, status, tags } = filter || {}
      if (params.searchTerm === '') {
        delete params.searchTerm
      }
      const requestBodyPayload = Object.assign(
        filter
          ? {
              delegateGroupIdentifier,
              delegateName,
              delegateType,
              description,
              hostName,
              status,
              tags
            }
          : {},
        {
          filterType: 'Delegate'
        }
      )
      const sanitizedFilterRequest = removeNullAndEmpty(requestBodyPayload)
      try {
        setDelegateFetchError(undefined)
        const delegateResponse = await fetchDelegates(sanitizedFilterRequest, { queryParams: params })
        const delGroups = delegateResponse?.resource?.delegateGroupDetails || []
        if (delGroups) {
          delGroups.forEach((delegateGroup: DelegateGroupDetails) => {
            const delName = `${get(delegateGroup, 'groupName', '')} ${getString('delegate.instancesCount', {
              count: delegateGroup?.delegateInstanceDetails?.length,
              total: ''
            })}`
            set(delegateGroup, 'delegateName', delName)
          })
          setDelegateGroups(delGroups)
        }
      } catch (e) {
        if (shouldShowError(e)) {
          showError(e.data?.message || e.message)
        }
        setDelegateFetchError(e)
      } finally {
        setShowDelegateLoader(false)
      }
    },
    [fetchDelegates]
  )

  // Add polling
  useEffect(() => {
    let timeoutId = 0

    if (!isFetchingDelegates && !delegateFetchError && !isFilterOpen) {
      timeoutId = window.setTimeout(() => {
        setShowDelegateLoader(false)
        refetchDelegates({ ...queryParams, searchTerm }, appliedFilter?.filterProperties)
      }, POLLING_INTERVAL)
    }

    return () => {
      clearTimeout(timeoutId)
    }
  }, [isFetchingDelegates, delegateFetchError, refetchDelegates, isFilterOpen])

  const formatPayload = (data: FilterDataInterface<DelegateFilterProperties, FilterInterface>, isUpdate: boolean) => {
    const {
      metadata: { name: _name, filterVisibility, identifier },
      formValues: { delegateGroupIdentifier, delegateName, delegateType, description, hostName, status, tags }
    } = data
    return {
      name: _name,
      identifier: isUpdate ? identifier : StringUtils.getIdentifierFromName(_name),
      projectIdentifier,
      orgIdentifier,
      filterVisibility: filterVisibility,
      filterProperties: {
        filterType: 'Delegate',
        delegateGroupIdentifier,
        delegateName,
        delegateType,
        description,
        hostName,
        status,
        tags
      } as DelegateFilterProperties
    }
  }

  const handleSaveOrUpdate = async (
    isUpdate: boolean,
    data: FilterDataInterface<DelegateFilterProperties, FilterInterface>
  ): Promise<void> => {
    await setIsRefreshingFilters(true)
    const requestBodyPayload = formatPayload(data, isUpdate)
    const saveOrUpdateHandler = filterRef.current?.saveOrUpdateFilterHandler
    if (saveOrUpdateHandler && typeof saveOrUpdateHandler === 'function') {
      const updatedFilter = await saveOrUpdateHandler(isUpdate, requestBodyPayload)
      setAppliedFilter(updatedFilter)
    }
    await refetchFilterList()
    setIsRefreshingFilters(false)
  }

  const handleDelete = async (identifier: string): Promise<void> => {
    setIsRefreshingFilters(true)
    const deleteHandler = filterRef.current?.deleteFilterHandler
    if (deleteHandler && typeof deleteFilter === 'function') {
      await deleteHandler(identifier)
    }
    if (identifier === appliedFilter?.identifier) {
      reset()
    }
    await refetchFilterList()
    setIsRefreshingFilters(false)
  }

  const unsavedFilter = {
    name: UNSAVED_FILTER,
    identifier: StringUtils.getIdentifierFromName(UNSAVED_FILTER)
  }

  const handleFilterClick = (identifier: string): void => {
    if (identifier !== unsavedFilter.identifier) {
      setAppliedFilter(getFilterByIdentifier(identifier))
    }
  }

  const DelegateForm = (): React.ReactElement => {
    return (
      <>
        <FormInput.Text name={'delegateName'} label={getString('delegate.delegateName')} key={'delegateName'} />
        <FormInput.Text name={'delegateType'} label={getString('delegate.delegateType')} key={'delegateType'} />
        <FormInput.Text name={'description'} label={getString('description')} key={'description'} />
        <FormInput.Text name={'hostName'} label={getString('delegate.hostName')} key={'hostName'} />
        <FormInput.Select
          items={getStatusSelectOptions(statusTypes)}
          name={'status'}
          label={getString('status')}
          key={'status'}
        />
        <FormInput.Text
          name={'delegateGroupIdentifier'}
          label={getString('delegates.delegateIdentifier')}
          key={'delegateGroupIdentifier'}
        />
        <FormInput.KVTagInput name="tags" label={getString('tagsLabel')} key="tags" />
      </>
    )
  }

  const { mutate: createFilter } = usePostFilter({
    queryParams: {
      accountIdentifier: accountId
    }
  })

  const { mutate: updateFilter } = useUpdateFilter({
    queryParams: {
      accountIdentifier: accountId
    }
  })

  const { mutate: deleteFilter } = useDeleteFilter({
    queryParams: {
      accountIdentifier: accountId,
      projectIdentifier,
      orgIdentifier,
      type: 'Delegate'
    }
  })

  const [openFilterDrawer, hideFilterDrawer] = useModalHook(() => {
    const onFilterApply = (formData: Record<string, any>) => {
      if (!isObjectEmpty(formData)) {
        const filterFromFormData = { ...formData }
        const updatedQueryParams = {
          ...queryParams,
          searchTerm,
          pageIndex: 0
        }
        setShowDelegateLoader(true)
        refetchDelegates(updatedQueryParams, filterFromFormData)
        setAppliedFilter({ ...unsavedFilter, filterProperties: filterFromFormData })
        setPage(0)
        hideFilterDrawer()
        setIsFilterOpen(false)
      } else {
        showError(getString('filters.invalidCriteria'))
      }
    }

    const { delegateName, delegateGroupIdentifier, delegateType, description, hostName, status, tags } =
      (appliedFilter?.filterProperties as any) || {}
    const { name = '', filterVisibility } = appliedFilter || {}
    return isFetchingFilters ? (
      <PageSpinner />
    ) : (
      <Filter<DelegateFilterProperties, FilterDTO>
        onApply={onFilterApply}
        onClose={() => {
          hideFilterDrawer()
          setIsFilterOpen(false)
          refetchFilterList()
        }}
        filters={filters}
        initialFilter={{
          formValues: {
            delegateName,
            delegateGroupIdentifier,
            delegateType,
            description,
            hostName,
            status,
            tags
          },
          metadata: {
            name,
            filterVisibility: filterVisibility,
            identifier: appliedFilter?.identifier || '',
            filterProperties: {}
          }
        }}
        onSaveOrUpdate={handleSaveOrUpdate}
        onDelete={handleDelete}
        onFilterSelect={handleFilterClick}
        isRefreshingFilters={isRefreshingFilters}
        formFields={<DelegateForm />}
        onValidate={(values: Partial<DelegateFilterProperties>): FormikErrors<Partial<DelegateFilterProperties>> => {
          const errors: FormikErrors<{ description?: MultiSelectOption[] }> = {}
          if (values.description === '') {
            errors.description = ''
          }
          return errors
        }}
        dataSvcConfig={
          new Map<CrudOperation, (...rest: any[]) => Promise<any>>([
            ['ADD', createFilter],
            ['UPDATE', updateFilter],
            ['DELETE', deleteFilter]
          ])
        }
        onSuccessfulCrudOperation={refetchFilterList}
        ref={filterRef}
        onClear={reset}
      />
    )
  }, [isRefreshingFilters, filters, appliedFilter, searchTerm, queryParams])

  const reset = (): void => {
    setShowDelegateLoader(true)
    refetchDelegates(queryParams)
    setAppliedFilter(undefined)
    setDelegateFetchError(undefined)
  }

  const debouncedDelegateSearch = useCallback(
    debounce((query: string): void => {
      /* For a non-empty query string, always start from first page(index 0) */
      const updatedQueryParams = {
        ...queryParams,
        searchTerm: query,
        pageIndex: 0
      }
      setShowDelegateLoader(true)
      refetchDelegates(updatedQueryParams, appliedFilter?.filterProperties)
    }, 500),
    [refetchDelegates, appliedFilter?.filterProperties]
  )

  const getFilterByIdentifier = (identifier: string): FilterDTO | undefined =>
    /* istanbul ignore if */
    filters?.find((filter: FilterDTO) => filter.identifier?.toLowerCase() === identifier.toLowerCase())

  const handleFilterSelection = (
    option: SelectOption,
    event?: React.SyntheticEvent<HTMLElement, Event> | undefined
  ): void => {
    event?.stopPropagation()
    event?.preventDefault()
    /* istanbul ignore else */
    if (option.value) {
      const selectedFilter = getFilterByIdentifier(option.value?.toString())
      setAppliedFilter(selectedFilter)
      const updatedQueryParams = {
        ...queryParams,
        searchTerm,
        pageIndex: 0
      }
      setShowDelegateLoader(true)
      refetchDelegates(updatedQueryParams, selectedFilter?.filterProperties)
    } else {
      reset()
    }
  }

  const fieldToLabelMapping = new Map<string, string>()
  fieldToLabelMapping.set('delegateGroupIdentifier', getString('delegates.delegateIdentifier'))
  fieldToLabelMapping.set('delegateName', getString('delegate.delegateName'))
  fieldToLabelMapping.set('delegateType', getString('delegate.delegateType'))
  fieldToLabelMapping.set('description', getString('description'))
  fieldToLabelMapping.set('hostName', getString('delegate.hostName'))
  fieldToLabelMapping.set('status', getString('status'))
  fieldToLabelMapping.set('tags', getString('delegate.delegateTags'))

  const permissionRequestNewDelegate = {
    resourceScope: {
      accountIdentifier: accountId,
      orgIdentifier,
      projectIdentifier
    },
    permission: PermissionIdentifier.UPDATE_DELEGATE,
    resource: {
      resourceType: ResourceType.DELEGATE
    }
  }

  const hideHeader = delegateGroups.length === 0 && !appliedFilter && !searchTerm

  const newDelegateBtn = (
    <RbacButton
      intent="primary"
      text={hideHeader ? getString('delegates.createDelegate') : getString('delegates.newDelegate')}
      icon="plus"
      permission={permissionRequestNewDelegate}
      onClick={() => openDelegateModal()}
      id="newDelegateBtn"
      data-testid="newDelegateButton"
    />
  )

  return (
    <Container height="100%">
      <Dialog
        isOpen={!!troubleshoterOpen}
        enforceFocus={false}
        style={{ width: '680px', height: '100%' }}
        onClose={() => setOpenTroubleshoter(undefined)}
      >
        <DelegateInstallationError showDelegateInstalledMessage={false} />
      </Dialog>
      {!hideHeader && (
        <Layout.Horizontal className={css.header}>
          {newDelegateBtn}
          <FlexExpander />
          <Layout.Horizontal spacing="xsmall">
            <ExpandingSearchInput
              alwaysExpanded
              width={250}
              placeholder={getString('search')}
              throttle={200}
              onChange={text => {
                debouncedDelegateSearch(text)
                setSearchTerm(text.trim())
              }}
              className={css.search}
            />
            <FilterSelector<FilterDTO>
              appliedFilter={appliedFilter}
              filters={filters}
              onFilterBtnClick={() => {
                openFilterDrawer()
                setIsFilterOpen(true)
              }}
              onFilterSelect={handleFilterSelection}
              fieldToLabelMapping={fieldToLabelMapping}
              filterWithValidFields={removeNullAndEmpty(
                pick(flattenObject(appliedFilter?.filterProperties || {}), ...fieldToLabelMapping.keys())
              )}
            />
          </Layout.Horizontal>
        </Layout.Horizontal>
      )}
      <Layout.Vertical className={css.listBody}>
        {showDelegateLoader ? (
          <Container style={fullSizeContentStyle}>
            <ContainerSpinner />
          </Container>
        ) : delegateFetchError && shouldShowError(delegateFetchError) ? (
          <div style={{ paddingTop: '200px' }}>
            <PageError
              message={(delegateFetchError?.data as Error)?.message || delegateFetchError?.message}
              onClick={(e: React.MouseEvent<Element, MouseEvent>) => {
                e.preventDefault()
                e.stopPropagation()
                reset()
              }}
            />
          </div>
        ) : (
          <Container className={css.delegateListContainer}>
            {delegateGroups.length ? (
              <Container width="100%">
                <DelegateListingHeader />
                {delegateGroups.map((delegate: DelegateGroupDetails) => (
                  <DelegateListingItem
                    key={delegate.delegateGroupIdentifier}
                    delegate={delegate}
                    setOpenTroubleshoter={setOpenTroubleshoter}
                  />
                ))}
              </Container>
            ) : (
              <div className={css.emptyStateContainer}>
                <img src={DelegatesEmptyState} />
                <div className={css.emptyStateText}>
                  {projectIdentifier
                    ? getString('delegates.noDelegatesInProject')
                    : orgIdentifier
                    ? getString('delegates.noDelegatesInOrganization')
                    : getString('delegates.noDelegatesInAccount')}
                </div>
                {hideHeader && newDelegateBtn}
              </div>
            )}
          </Container>
        )}
      </Layout.Vertical>
    </Container>
  )
}
export default DelegateListing
