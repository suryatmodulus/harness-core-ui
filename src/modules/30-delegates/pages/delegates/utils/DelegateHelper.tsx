/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { useStrings } from 'framework/strings'
import { DelegateTypes } from '@delegates/constants'

export const GetDelegateTitleTextByType = (type: string): string => {
  const { getString } = useStrings()

  switch (type) {
    case DelegateTypes.KUBERNETES_CLUSTER:
      return getString('kubernetesText')
    default:
      /* istanbul ignore next */
      return ''
  }
}
