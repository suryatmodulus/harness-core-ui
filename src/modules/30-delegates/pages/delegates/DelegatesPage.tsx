/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useEffect, useState } from 'react'
import { useParams } from 'react-router-dom'
import { TabNavigation } from '@wings-software/uicore'
import { NGBreadcrumbs } from '@common/components/NGBreadcrumbs/NGBreadcrumbs'
import { Page } from '@common/exports'
import ScopedTitle from '@common/components/Title/ScopedTitle'
import { Scope } from '@common/interfaces/SecretsInterface'
import routes from '@common/RouteDefinitions'
import type { ProjectPathProps, PipelineType } from '@common/interfaces/RouteInterfaces'
import { getLinkForAccountResources } from '@common/utils/BreadcrumbUtils'
import type { GetDelegateGroupsNGV2WithFilterQueryParams } from 'services/portal'
import { useListDelegateConfigsNgV2WithFilter } from 'services/cd-ng'
import type { DelegateProfileDetailsNg } from 'services/cd-ng'
import { useStrings } from 'framework/strings'

const DelegatesPage: React.FC = ({ children }) => {
  const params = useParams<PipelineType<ProjectPathProps>>()
  const { accountId, orgIdentifier, projectIdentifier, module } = params
  const { getString } = useStrings()
  const [profiles, setProfiles] = useState<DelegateProfileDetailsNg[]>([])

  const { mutate: getDelegateProfiles } = useListDelegateConfigsNgV2WithFilter({
    accountId,
    queryParams: { orgId: orgIdentifier, projectId: projectIdentifier }
  })

  const getDelegates = async () => {
    const delProfilesResponse = await getDelegateProfiles(
      {
        filterType: 'DelegateProfile'
      },
      {
        queryParams: {
          accountId,
          module,
          pageSize: 10,
          searchTerm: ''
        } as GetDelegateGroupsNGV2WithFilterQueryParams
      }
    )
    const resource = delProfilesResponse?.resource as any
    setProfiles(resource?.response || [])
  }

  useEffect(() => {
    getDelegates()
  }, [getDelegateProfiles])

  const links = [
    {
      label: getString('delegate.delegates'),
      to: routes.toDelegateList({ accountId, orgIdentifier, projectIdentifier, module })
    }
  ]
  if (profiles.length > 0) {
    links.push({
      label: getString('delegate.delegateConfigurations'),
      to: routes.toDelegateConfigs({ accountId, orgIdentifier, projectIdentifier, module })
    })
  }
  links.push({
    label: getString('common.tokens'),
    to: routes.toDelegateTokens({ accountId, orgIdentifier, projectIdentifier, module })
  })
  return (
    <>
      <Page.Header
        breadcrumbs={
          <NGBreadcrumbs
            links={getLinkForAccountResources({ accountId, orgIdentifier, projectIdentifier, getString })}
          />
        }
        title={
          <ScopedTitle
            title={{
              [Scope.PROJECT]: getString('delegate.delegates'),
              [Scope.ORG]: getString('delegates.delegatesTitle'),
              [Scope.ACCOUNT]: getString('delegates.delegatesTitle')
            }}
          />
        }
        toolbar={<TabNavigation size={'small'} links={links} />}
      />
      <Page.Body>{children}</Page.Body>
    </>
  )
}

export default DelegatesPage
