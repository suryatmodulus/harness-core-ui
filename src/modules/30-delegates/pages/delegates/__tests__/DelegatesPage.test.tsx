/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { render } from '@testing-library/react'
import { TestWrapper } from '@common/utils/testUtils'
import DelegatesPage from '../DelegatesPage'
import DelegateListing from '../DelegateListing'
import ProfileMock from './ProfilesMock'

const mockGetCallFunction = jest.fn()
jest.mock('services/portal', () => ({
  useGetDelegateGroupsNGV2WithFilter: jest.fn().mockImplementation(args => {
    mockGetCallFunction(args)
    return []
  })
}))

jest.mock('services/cd-ng', () => ({
  useListDelegateConfigsNgV2WithFilter: jest.fn().mockImplementation(args => {
    mockGetCallFunction(args)
    return {
      mutate: jest.fn().mockImplementation(() => ProfileMock),
      refetch: jest.fn(),
      error: null,
      loading: false
    }
  }),
  useGetFilterList: jest.fn().mockImplementation(() => {
    return {
      mutate: jest.fn()
    }
  }),
  usePostFilter: jest.fn().mockImplementation(() => {
    return {
      mutate: jest.fn()
    }
  }),
  useUpdateFilter: jest.fn().mockImplementation(() => {
    return {
      mutate: jest.fn()
    }
  }),
  useDeleteFilter: jest.fn().mockImplementation(() => {
    return {
      mutate: jest.fn()
    }
  })
}))

jest.mock('@common/components/YAMLBuilder/YamlBuilder')
describe('Delegates Page', () => {
  test('render data', () => {
    const { container } = render(
      <TestWrapper path="/account/:accountId/resources/delegates" pathParams={{ accountId: 'dummy' }}>
        <DelegatesPage>
          <DelegateListing />
        </DelegatesPage>
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })
})
