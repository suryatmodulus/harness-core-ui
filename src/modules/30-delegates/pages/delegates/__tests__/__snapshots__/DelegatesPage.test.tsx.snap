// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Delegates Page render data 1`] = `
<div>
  <div
    class="Layout--horizontal StyledProps--main PageHeader--container PageHeader--standard StyledProps--flex StyledProps--padding-left-xlarge StyledProps--padding-right-xlarge StyledProps--background-white"
  >
    <div
      class="Layout--vertical StyledProps--main"
    >
      <div
        class="PageHeader--breadcrumbsDiv"
      >
        <div
          class="Layout--horizontal StyledProps--main Breadcrumbs--breadcrumbs StyledProps--flex StyledProps--flex-align-center-center StyledProps--flex-justifyContent-flex-start"
        >
          <div
            class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-align-center-center StyledProps--flex-justifyContent-flex-start"
          >
            <a
              class="Breadcrumbs--breadcrumb"
              href="/account/dummy/settings"
            >
              <span
                class="bp3-icon bp3-icon-cog StyledProps--main StyledProps--padding-right-xsmall StyledProps--color StyledProps--color-primary7"
                icon="cog"
              >
                <svg
                  data-icon="cog"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <desc>
                    cog
                  </desc>
                  <path
                    d="M15.19 6.39h-1.85c-.11-.37-.27-.71-.45-1.04l1.36-1.36c.31-.31.31-.82 0-1.13l-1.13-1.13a.803.803 0 00-1.13 0l-1.36 1.36c-.33-.17-.67-.33-1.04-.44V.79c0-.44-.36-.8-.8-.8h-1.6c-.44 0-.8.36-.8.8v1.86c-.39.12-.75.28-1.1.47l-1.3-1.3c-.3-.3-.79-.3-1.09 0L1.82 2.91c-.3.3-.3.79 0 1.09l1.3 1.3c-.2.34-.36.7-.48 1.09H.79c-.44 0-.8.36-.8.8v1.6c0 .44.36.8.8.8h1.85c.11.37.27.71.45 1.04l-1.36 1.36c-.31.31-.31.82 0 1.13l1.13 1.13c.31.31.82.31 1.13 0l1.36-1.36c.33.18.67.33 1.04.44v1.86c0 .44.36.8.8.8h1.6c.44 0 .8-.36.8-.8v-1.86c.39-.12.75-.28 1.1-.47l1.3 1.3c.3.3.79.3 1.09 0l1.09-1.09c.3-.3.3-.79 0-1.09l-1.3-1.3c.19-.35.36-.71.48-1.1h1.85c.44 0 .8-.36.8-.8v-1.6a.816.816 0 00-.81-.79zm-7.2 4.6c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              <p
                class="StyledProps--font StyledProps--main StyledProps--intent-primary StyledProps--font-size-small"
              >
                common.accountSettings
              </p>
            </a>
            <span
              class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey500 StyledProps--padding-right-xsmall StyledProps--padding-left-xsmall"
              data-icon="main-chevron-right"
            >
              <svg
                height="8"
                viewBox="0 0 8 14"
                width="8"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7.7 6.3l-6-6C1.5.1 1.3 0 1 0 .4 0 0 .4 0 1c0 .3.1.5.3.7L5.6 7 .3 12.3c-.2.2-.3.4-.3.7 0 .6.4 1 1 1 .3 0 .5-.1.7-.3l6-6c.2-.2.3-.4.3-.7s-.1-.5-.3-.7z"
                  fill="#4d5969"
                />
              </svg>
            </span>
          </div>
          <div
            class="Layout--horizontal StyledProps--main StyledProps--flex StyledProps--flex-align-center-center StyledProps--flex-justifyContent-flex-start"
          >
            <a
              class="Breadcrumbs--breadcrumb"
              href="/account/dummy/settings/resources"
            >
              <p
                class="StyledProps--font StyledProps--main StyledProps--intent-primary StyledProps--font-size-small"
              >
                common.accountResources
              </p>
            </a>
            <span
              class="bp3-icon StyledProps--main StyledProps--color StyledProps--color-grey500 StyledProps--padding-right-xsmall StyledProps--padding-left-xsmall"
              data-icon="main-chevron-right"
            >
              <svg
                height="8"
                viewBox="0 0 8 14"
                width="8"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7.7 6.3l-6-6C1.5.1 1.3 0 1 0 .4 0 0 .4 0 1c0 .3.1.5.3.7L5.6 7 .3 12.3c-.2.2-.3.4-.3.7 0 .6.4 1 1 1 .3 0 .5-.1.7-.3l6-6c.2-.2.3-.4.3-.7s-.1-.5-.3-.7z"
                  fill="#4d5969"
                />
              </svg>
            </span>
          </div>
        </div>
      </div>
      <div
        class="Layout--horizontal StyledProps--main"
      >
        <p
          class="StyledProps--font StyledProps--main StyledProps--font-variation-h4"
        >
          account delegates.delegatesTitle
        </p>
      </div>
    </div>
    <div
      class="Layout--horizontal Layout--layout-spacing-small StyledProps--main TabNavigation--container"
    >
      <a
        class="TabNavigation--tags TabNavigation--small"
        href="/account/dummy/settings/resources/delegates/list"
      >
        delegate.delegates
      </a>
      <a
        class="TabNavigation--tags TabNavigation--small"
        href="/account/dummy/settings/resources/delegates/tokens"
      >
        common.tokens
      </a>
    </div>
  </div>
  <div
    class="StyledProps--font StyledProps--main PageBody--pageBody"
    data-name="page-body"
  >
    <div
      class="StyledProps--font StyledProps--main"
      style="height: 100%;"
    >
      <div
        class="Layout--vertical StyledProps--main listBody"
      >
        <div
          style="padding-top: 200px;"
        >
          <div
            class="StyledProps--font StyledProps--main StyledProps--flex StyledProps--flex-align-center-center"
            style="width: 100%; height: 100%;"
          >
            <div
              class="Layout--vertical Layout--layout-spacing-medium StyledProps--main"
              style="align-items: center; width: 500px;"
            >
              <span
                class="bp3-icon bp3-icon-error StyledProps--main StyledProps--color StyledProps--color-red500"
                icon="error"
              >
                <svg
                  data-icon="error"
                  height="32"
                  viewBox="0 0 20 20"
                  width="32"
                >
                  <desc>
                    error
                  </desc>
                  <path
                    d="M10 0C4.48 0 0 4.48 0 10s4.48 10 10 10 10-4.48 10-10S15.52 0 10 0zm1 16H9v-2h2v2zm0-3H9V4h2v9z"
                    fill-rule="evenodd"
                  />
                </svg>
              </span>
              <p
                class="StyledProps--font StyledProps--main StyledProps--font-align-center StyledProps--color StyledProps--color-red500"
              >
                fetchDelegates is not a function
              </p>
              <button
                class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--variation Button--variation-primary"
                type="button"
              >
                <span
                  class="bp3-button-text"
                >
                  Retry
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
